src/libutil/auth.hh

```diff
@@ -0,0 +1,144 @@
+#pragma once
+
+#include <optional>
+
+#include "types.hh"
+#include "config.hh"
+
+namespace nix::auth {
+
+enum struct AuthForwarding { Disabled, TrustedUsers, AllUsers };
+
+struct AuthSettings : Config
+{
+    Setting<Strings> authSources{
+        this,
+        {"builtin:nix", "builtin:netrc"},
+        "auth-sources",
+        R"(
+          A list of helper programs from which to obtain
+          authentication data for HTTP requests.  These helpers use
+          [the same protocol as Git's credential
+          helpers](https://git-scm.com/docs/gitcredentials#_custom_helpers),
+          so any Git credential helper can be used as an
+          authentication source.
+
+          Nix has the following builtin helpers:
+
+          * `builtin:nix`: Get authentication data from files in
+            `~/.local/share/nix/auth`. For example, the following sets
+            a username and password for `cache.example.org`:
+
+            ```
+            # cat <<EOF > ~/.local/share/nix/auth/my-cache
+            protocol=https
+            host=cache.example.org
+            username=alice
+            password=foobar
+            EOF
+            ```
+
+          Example: `builtin:nix` `git-credential-libsecret`
+        )"};
+
+    Setting<Path> netrcFile{
+        this, "", "netrc-file",
+        R"(
+          An absolute path to a `netrc` file. Nix will use the HTTP
+          authentication credentials in this file when trying to download from
+          a remote host through HTTP or HTTPS. Defaults to
+          `$NIX_CONF_DIR/netrc`.
+
+          The `netrc` file consists of a list of accounts in the following
+          format:
+
+              machine my-machine
+              login my-username
+              password my-password
+
+          For the exact syntax, see [the `curl`
+          documentation](https://ec.haxx.se/usingcurl-netrc.html).
+
+          > **Note**
+          >
+          > This must be an absolute path, and `~` is not resolved. For
+          > example, `~/.netrc` won't resolve to your home directory's
+          > `.netrc`.
+        )"};
+
+    Setting<bool> storeAuth{
+        this, false, "store-auth",
+        R"(
+          Whether to store user names and passwords using the
+          authentication sources configured in [`auth-sources`](#conf-auth-sources).
+        )"};
+
+    Setting<AuthForwarding> authForwarding{
+        this, AuthForwarding::TrustedUsers, "auth-forwarding",
+        R"(
+          Whether to forward authentication data to the Nix daemon. This setting can have the following values:
+
+          * `false`: Forwarding is disabled.
+          * `trusted-users`: Forwarding is only supported for [trusted users](#conf-trusted-users).
+          * `all-users`: Forwarding is supported for all users.
```

Comment:
What happens to the daemon's own config? Is it ignored completely or merged? If so, how?

Thought:

It's unfortunate that we can't seem to just enable this for all users by default. (Or am I wrong to think I shouldn't make myself trusted user? Cachix has certainly normalized it...)

*Receiving* auth data seems pretty harmless on the face of it, but I suppose attack vectors may involve:
- Implementation errors in our protocol clients
- Bugs in remotes that cause them to misbehave while we trust them
- Remotes that behave differently depending on user name

I wonder if there's any mitigations we could apply? Perhaps a scenario like this:

- a cache `https://foo` exists
- `https://foo/nix-cache-info` has a header that declares that its happy-path behavior is not auth-dependent (e.g. doesn't select a cache based on username)
- `nix daemon` has `trusted-substituters = https://foo`
- trusted substituter + header allow credential forwarding

Also the retrieval of content addressed paths seems pretty safe, because misbehaving remotes are not much of an issue.

Reply:
> Receiving auth data seems pretty harmless on the face of it

Yeah, this is not harmless because we do indeed have substituters that return different data depending on the auth data, e.g. cache.flakehub.com (which is based on Attic) does that.

> https://foo/nix-cache-info has a header that declares that its happy-path behavior is not auth-dependent (e.g. doesn't select a cache based on username)

Yeah that sounds like a good idea. Probably for a future PR though. There's a bit of a chicken/egg problem with that since to fetch `nix-cache-info`, we may already need to know the auth data.

Reply:
> What happens to the daemon's own config?

The daemon will try its own authentication sources *and* those forwarded by the client.